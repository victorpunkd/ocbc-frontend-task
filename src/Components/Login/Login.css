.loginPageContainer {
  width: 100%;
  height: 100%;
  position: relative;
}

.loginForm {
  margin: 0;
  position: relative;
  top: 50%;
  transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  text-align: center;
}

.headingText {
  font-size: 25px;
  font-weight: 700;
  color: #dd54a5;
}

.subHeadingText {
  font-size: 20px;
  color: #757575;
}

.formControls {
  margin-top: 5%;
}
