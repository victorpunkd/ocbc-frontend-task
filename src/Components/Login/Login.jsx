import React, { useEffect, useState } from "react";
import axios from "axios";
import { useNavigate } from "react-router-dom";
import "./Login.css";
import CustomButton from "../../CommonComponents/CustomButton/CustomButton";
import CustomSnackBar from "../../CommonComponents/CustomSnackBar/CustomSnackBar";
import TextField from "@mui/material/TextField";
import LoginIcon from "@mui/icons-material/Login";
import CustomBackdrop from "../../CommonComponents/CustomBackdrop/CustomBackdrop";

const Login = () => {
  const navigate = useNavigate();
  const [password, setPassword] = useState("");
  const [userName, setUserName] = useState("");
  const [isBackdropOpen, setIsBackdropOpen] = useState(false);
  const [isSnackbarOpen, setIsSnackbarOpen] = useState(false);
  const [snackBarSeverity, setSnackBarSeverity] = useState("");
  const [snackBarMessage, setSnackBarMessage] = useState("");
  const [errorState, setErrorState] = useState({});

  const handleChange = (type, event) => {
    if (type === "userName") {
      setUserName(event.target.value);
    } else if (type === "password") {
      setPassword(event.target.value);
    }
  };

  const validateDetails = () => {
    const errors = {};
    if (!userName) {
      errors["userName"] = "Required";
    }
    if (!password) {
      errors["password"] = "Required";
    }
    return errors;
  };

  const handleLoginClick = () => {
    const errors = validateDetails();
    setErrorState(errors);
    if (Object.keys(errors).length !== 0) {
      setErrorState(errors);
      return;
    } else {
      setErrorState({});
    }
    loginUser();
  };

  const handleBackEndError = (error) => {
    setIsSnackbarOpen(true);
    setSnackBarMessage(error?.response?.data?.description || "");
    setSnackBarSeverity("error");
  };

  useEffect(() => {
    sessionStorage.removeItem("token");
  }, []);

  const loginUser = () => {
    setIsBackdropOpen(true);
    axios
      .post(`${process.env.REACT_APP_BASE_URL}authenticate/login`, {
        username: userName,
        password: password,
      })
      .then((response) => {
        if (
          response.data &&
          response.data.status &&
          response.data.status === "success"
        ) {
          if (response.data.token) {
            sessionStorage.setItem("token", response.data.token);
            navigate("/dashboard");
          }
        }
      })
      .catch((error) => {
        if (error.response.data && error.response.data.status) {
          handleBackEndError(error);
        }
      })
      .finally(() => {
        setIsBackdropOpen(false);
      });
  };

  return (
    <div className="loginPageContainer">
      <CustomSnackBar
        open={isSnackbarOpen}
        severity={snackBarSeverity}
        message={snackBarMessage}
        handleClose={() => setIsSnackbarOpen(false)}
      />
      <CustomBackdrop
        open={isBackdropOpen}
        handleClose={() => setIsBackdropOpen(false)}
      />
      <div className="loginForm">
        <div className="headingText">Hi there!</div>
        <div className="subHeadingText">Login with your credentials</div>
        <div style={{ marginTop: "3%" }}>
          <div className="formControls">
            <TextField
              label="Username"
              id="username"
              error={errorState["userName"]}
              fullWidth
              variant="outlined"
              className="textField"
              value={userName}
              onChange={(event) => handleChange("userName", event)}
            />
          </div>
          <div className="formControls">
            <TextField
              label="Password"
              id="password"
              fullWidth
              className="textField"
              error={errorState["password"]}
              type={"password"}
              variant="outlined"
              value={password}
              onChange={(event) => handleChange("password", event)}
            />
          </div>
        </div>
        <div style={{ marginTop: "5%" }}>
          <CustomButton
            id="loginButton"
            buttonText="Login"
            buttonClick={handleLoginClick}
            icon={<LoginIcon />}
          />
        </div>
      </div>
    </div>
  );
};

export default Login;
