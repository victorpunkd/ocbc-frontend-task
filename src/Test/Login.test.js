import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Login from "../Components/Login/Login";

configure({ adapter: new Adapter() });

const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe("Login", () => {
  it("snapshot renders", () => {
    const component = renderer.create(<Login />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should render a username, password, input tag and loginButton button", () => {
    const wrapper = shallow(<Login />);
    expect(wrapper.find("#username").exists()).toBe(true);
    expect(wrapper.find("#password").exists()).toBe(true);
    expect(wrapper.find("#loginButton").exists()).toBe(true);
  });

  it("Should simulate username onChange event", () => {
    const wrapper = shallow(<Login />);
    expect(
      wrapper.find("#username").simulate("change", {
        target: {
          value: "test",
        },
      })
    );

    expect(wrapper.find("#username").prop("value")).toBe("test");
  });

  it("Should simulate password onChange event", () => {
    const wrapper = shallow(<Login />);
    expect(
      wrapper.find("#password").simulate("change", {
        target: {
          value: "test",
        },
      })
    );

    expect(wrapper.find("#password").prop("value")).toBe("test");
  });

  it("Should simulate password onChange event", () => {
    const wrapper = shallow(<Login />);
    expect(
      wrapper.find("#password").simulate("change", {
        target: {
          value: "test",
        },
      })
    );

    expect(wrapper.find("#password").prop("value")).toBe("test");
  });

  it("Should simulate login onClick event", () => {
    const wrapper = shallow(<Login />);
    wrapper.find("#password").simulate("change", {
      target: {
        value: "test",
      },
    });
    wrapper.find("#username").simulate("change", {
      target: {
        value: "test",
      },
    });
    wrapper.find("#loginButton").props().buttonClick("");
    expect(wrapper.find("#loginButton").prop("buttonText")).toBe("Login");
  });
});
