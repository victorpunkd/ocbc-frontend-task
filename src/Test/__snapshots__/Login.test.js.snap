// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot renders 1`] = `
<div
  className="loginPageContainer"
>
  <div>
    <div
      aria-hidden={true}
      className="MuiBackdrop-root css-1ji7omv-MuiBackdrop-root"
      onClick={[Function]}
      style={
        Object {
          "opacity": 0,
          "visibility": "hidden",
        }
      }
    >
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorInherit css-62e83j-MuiCircularProgress-root"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className="loginForm"
  >
    <div
      className="headingText"
    >
      Hi there!
    </div>
    <div
      className="subHeadingText"
    >
      Login with your credentials
    </div>
    <div
      style={
        Object {
          "marginTop": "3%",
        }
      }
    >
      <div
        className="formControls"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root textField css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor="username"
            id="username-label"
          >
            Username
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id="username"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-dhh0px"
              >
                <span>
                  Username
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        className="formControls"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root textField css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor="password"
            id="password-label"
          >
            Password
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              id="password"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="password"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-dhh0px"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
    <div
      style={
        Object {
          "marginTop": "5%",
        }
      }
    >
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root makeStyles-buttonStyle-1 css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        id="loginButton"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        Login
        <span
          className="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="LoginIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 7 9.6 8.4l2.6 2.6H2v2h10.2l-2.6 2.6L11 17l5-5-5-5zm9 12h-8v2h8c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-8v2h8v14z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
